newton/utils/contact_sensor.py

```diff
@@ -0,0 +1,397 @@
+            hi = mid
+    return lo
+
+
+@wp.kernel
+def select_aggregate_net_force(
+    num_contacts: wp.array(dtype=wp.int32),
+    sp_sorted: wp.array(dtype=wp.vec2i),
+    num_sp: int,
+    sp_ep: wp.array(dtype=wp.vec2i),
+    sp_ep_offset: wp.array(dtype=wp.int32),
+    sp_ep_count: wp.array(dtype=wp.int32),
+    contact_pair: wp.array(dtype=wp.vec2i),
+    contact_normal: wp.array(dtype=wp.vec3f),
+    contact_force: wp.array(dtype=wp.float32),
+    # output
+    net_force: wp.array(dtype=wp.vec3),
+):
+    ncon = num_contacts[0]
+    n_blocks = (ncon + NUM_THREADS - 1) // NUM_THREADS
+    for b in range(n_blocks):
+        con_idx = wp.tid() + NUM_THREADS * b
+        if con_idx >= ncon:
+            return
+
+        pair = contact_pair[con_idx]
+
+        # Find the entity pairs
+        smin, smax = wp.min(pair[0], pair[1]), wp.max(pair[0], pair[1])
+
+        # add contribution for shape pair
+        normalized_pair = wp.vec2i(smin, smax)
+        sp_flip = normalized_pair[0] != pair[0]
+        sp_ord = bisect_shape_pairs(sp_sorted, num_sp, normalized_pair)
+
+        force = contact_force[con_idx] * contact_normal[con_idx]
+        if sp_ord < num_sp and sp_sorted[sp_ord] == normalized_pair:
+            # add the force to the pair's force accumulators
+            offset = sp_ep_offset[sp_ord]
+            for i in range(sp_ep_count[sp_ord]):
+                ep = sp_ep[offset + i]
+                force_acc, flip = ep[0], ep[1]
+                wp.atomic_add(net_force, force_acc, wp.where(sp_flip != flip, -force, force))
+
+        # add contribution for shape a and b
+        for i in range(2):
+            mono_sp = wp.vec2i(-1, pair[i])
+            mono_ord = bisect_shape_pairs(sp_sorted, num_sp, mono_sp)
+
+            # for shape vs all, only one accumulator is supported and flip is trivially true
+            if mono_ord < num_sp and sp_sorted[mono_ord] == mono_sp:
+                force_acc = sp_ep[sp_ep_offset[mono_ord]][0]
+                wp.atomic_add(net_force, force_acc, wp.where(bool(i), -force, force))
+
+
+def _lol_to_arrays(list_of_lists: list[list], dtype) -> tuple[wp.array, wp.array, wp.array]:
+    """Convert a list of lists to three warp arrays containing the values, offsets and counts.
+    Does nothing and returns None, None, None if the list is empty.
+    """
+    if not list_of_lists:
+        return None, None, None
+    a = wp.array([el for l in list_of_lists for el in l], dtype=dtype)
+    count_list = list(map(len, list_of_lists))
+    offset = wp.array(np.cumsum([0, *count_list[:-1]]), dtype=wp.int32)
+    count = wp.array(count_list, dtype=wp.int32)
+    return a, offset, count
+
+
+def convert_contact_info(
+    model: Model,
+    contact_info: ContactInfo,
+    solver: SolverBase | None = None,
+    contacts: Contacts | None = None,
+):
+    """Populate ContactInfo object from the solver or from the Contacts object."""
+    if solver is not None:
+        if isinstance(solver, MuJoCoSolver):
```

Comment:
I think there should be no dependency on a particular solver in this module.
We could add an abstract method to `SolverBase.update_contact_info(model, constacts, contact_info)` that does nothing by default and `MuJoCoSolver` implements that with what you have already. "update_newton_contacts" is not a good name I think because you are updating `ContactInfo`, not `newton.Contacts`. So consider naming it `update_contact_info`.